kubernetes/schema_resources_migrate.go

```diff
@@ -0,0 +1,149 @@
+	if !ok || len(js) == 0 {
+		return rawState, nil
+	}
+
+	jobSpec := js[0].(map[string]interface{})
+	t, ok := jobSpec["template"].([]interface{})
+	if !ok || len(t) == 0 {
+		return rawState, nil
+	}
+
+	template := t[0].(map[string]interface{})
+	ps, ok := template["spec"].([]interface{})
+	if !ok || len(ps) == 0 {
+		return rawState, nil
+	}
+
+	podSpec := ps[0].(map[string]interface{})
+	template["spec"] = []interface{}{upgradeContainers(podSpec)}
+
+	return rawState, nil
+}
+
+func upgradeTemplatePodSpecWithResourcesFieldV0(ctx context.Context, rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	s, ok := rawState["spec"].([]interface{})
+	if !ok || len(s) == 0 {
+		return rawState, nil
+	}
+
+	spec := s[0].(map[string]interface{})
+	t, ok := spec["template"].([]interface{})
+	if !ok || len(t) == 0 {
+		return rawState, nil
+	}
+
+	template := t[0].(map[string]interface{})
+	ps, ok := template["spec"].([]interface{})
+
+	if !ok || len(ps) == 0 {
+		return rawState, nil
+	}
+
+	podSpec := ps[0].(map[string]interface{})
+	template["spec"] = []interface{}{upgradeContainers(podSpec)}
+
+	return rawState, nil
+}
+
+func upgradePodSpecWithResourcesFieldV0(ctx context.Context, rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	s, ok := rawState["spec"].([]interface{})
+	if !ok || len(s) == 0 {
+		return rawState, nil
+	}
+
+	spec := s[0].(map[string]interface{})
+	rawState["spec"] = []interface{}{upgradeContainers(spec)}
+
+	return rawState, nil
+}
+
+func upgradeContainers(rawState map[string]interface{}) map[string]interface{} {
+	if initContainers, ok := rawState["init_container"].([]interface{}); ok && len(initContainers) > 0 {
+		for _, c := range initContainers {
+			initContainer := c.(map[string]interface{})
+			if r, ok := initContainer["resources"].([]interface{}); ok && len(r) > 0 {
+				resources := r[0].(map[string]interface{})
+				if req, ok := resources["requests"].([]interface{}); ok && len(req) > 0 {
+					requests := req[0].(map[string]interface{})
+					resources["requests"] = requests
+				}
+				if lim, ok := resources["limits"].([]interface{}); ok && len(lim) > 0 {
+					limits := lim[0].(map[string]interface{})
+					resources["limits"] = limits
+				}
+			}
+		}
+	}
+	if containers, ok := rawState["container"].([]interface{}); ok && len(containers) > 0 {
+		for _, c := range containers {
+			container := c.(map[string]interface{})
+			if r, ok := container["resources"].([]interface{}); ok && len(r) > 0 {
+				resources := r[0].(map[string]interface{})
+				if req, ok := resources["requests"].([]interface{}); ok && len(req) > 0 {
+					requests := req[0].(map[string]interface{})
+					resources["requests"] = requests
+				}
+				if lim, ok := resources["limits"].([]interface{}); ok && len(lim) > 0 {
+					limits := lim[0].(map[string]interface{})
+					resources["limits"] = limits
+				}
+			}
+		}
+	}
+	return rawState
+}
```

Comment:
I can tell by my testing that this function is doing what it's supposed to do, but I can't seem to figure out how it works. Maybe next year we can refactor it to be easier to follow. Specifically, I'm looking at how this function returns `rawState`... but `rawState` is never modified (as far as I can tell). A variable named `resources` is created and modified to contain the right information, but I have no idea how that makes its way into `rawState`. :thinking: 

To give you a clearer picture of what I mean, here's how I made my equivalent of this function for a StateUpgrader: https://github.com/hashicorp/terraform-provider-kubernetes/blob/74244d7f1902e3c586657d50e81ceb896a01e30c/kubernetes/resource_kubernetes_service_migrate.go#L179-L202

It does a a couple things:
1. Create a new object to hold the migrated state.
2. Descend into the old object, checking that certain fields are set, and copying them into the new object if they do.

I see how your function descends into the old object and checks fields too. And copies those fields into a new object... I just don't understand what happens after that, since the old object is returned.

Reply:
I'm finding I have this same confusion with every function that returns rawState instead of the new objects. :exploding_head:  I trust that it works. It's just not something I'm familiar with.

Reply:
Yeah it is confusing, we could rewrite it to return an entirely new object like you have done in your other PR. It works because maps are mutable, so even if you assign a map within a map to a new var you are still modifying the underlying map, not creating a new one. 🤯